src/k2/dto/FieldTypes.cpp

```diff
@@ -33,51 +33,97 @@ Copyright(c) 2020 Futurewei Cloud
 namespace k2 {
 namespace dto {
 
-static constexpr char TERM = '\0';
+static constexpr char ESCAPE = '\0';
+static constexpr char TERM = 0x01;
+static constexpr char ESCAPED_NULL = 0xFF;
 
 template <> FieldType TToFieldType<String>() { return FieldType::STRING; }
 template <> FieldType TToFieldType<uint32_t>() { return FieldType::UINT32T; }
 
 // All conversion assume ascending ordering
 
-template <> String FieldToKeyString<String>(const String& field) {
-    // Sizes will not match if there are exta null bytes
-    K2ASSERT(field.size() == strlen(field.c_str()), "String has null bytes");
+static String escapeKeyString(const String& field) {
```

Comment:
For efficiency, i wonder if it also makes sense to provide a concat & escape method, where you can perform the escaping and terminator-based concatenation in one allocation.

Reply:
Also, i think it'd be better if this method is a noop by itself if the string doesn't need escaping. That way you don't have to check for noop before calling it like you do below. I think one way to do this is via a signature
bool escape(const String& in, String& out). If no-op, we don't set the output.
Another way is to use
String escape(String&& in). If no-op, we set the output to the input by move.

Reply:
I combined them to one function to avoid the copy and make it a bit cleaner.